actors/ppo_atari_pz.py

```diff
@@ -0,0 +1,1085 @@
+                trials_id_and_params=[
+                    (f"{run_session.run_id}_{iter_idx}_{trial_idx}", create_trial_params(trial_idx, iter_idx))
+                    for trial_idx in range(self._cfg.epoch_num_trials)
+                ],
+                sample_producer_impl=self.trial_sample_sequences_producer_impl,
+                num_parallel_trials=self._cfg.num_parallel_trials,
+            ):
+                # Collect the rollout
+                (trial_observation, trial_action, trial_reward, trial_done) = sample
+
+                observations.extend(trial_observation)
+                actions.extend(trial_action)
+                rewards.extend(trial_reward)
+                dones.extend(trial_done)
+
+                episode_rewards.append(torch.vstack(trial_reward).sum())
+                episode_lens.append(torch.tensor(len(trial_action), dtype=torch.float32))
+
+                # Publish the newly trained version every 100 steps
+                if len(actions) >= self._cfg.num_steps * self._cfg.epoch_num_trials + 1:
+                    num_updates += 1
+                    total_steps += self._cfg.num_steps * self._cfg.epoch_num_trials
+                    # Update model parameters
+                    policy_loss, value_loss = await self.train_step(
+                        observations=observations, rewards=rewards, actions=actions, dones=dones
+                    )
+
+                    # Reset the data storage
+                    observations = []
+                    actions = []
+                    rewards = []
+                    dones = []
+                    if iter_idx % 1 == 0:
+                        # Compute average rewards for last 100 episodes
+                        avg_rewards = await self.compute_average_reward(episode_rewards)
+                        avg_lens = await self.compute_average_reward(episode_lens) / self._environment_specs.num_players
+                        log.info(f"epoch #{iter_idx + 1}/{self._cfg.num_iter}| avg. len: {avg_lens:0.2f}]")
+
+                        run_session.log_metrics(
+                            model_version_number=version_info["version_number"],
+                            policy_loss=policy_loss.item(),
+                            value_loss=value_loss.item(),
+                            avg_rewards=avg_rewards.item(),
+                            avg_lens=avg_lens.item(),
+                            num_steps=total_steps,
+                            num_updates=num_updates,
+                        )
+
+                    # Publish the newly updated model
+                    self.model.iter_idx = iter_idx
+                    if num_updates % 50 == 0:
+                        version_info = await run_session.model_registry.publish_version(self.model, archived=True)
+                    else:
+                        version_info = await run_session.model_registry.publish_version(self.model)
+                    self.model.network.to(self._device)
+
```

Comment:
Shouldn't you run evaluation trials at the end of each epoc with the current player against an old version of itself to monitor progress ?

Reply:
I think it is a good idea for the next version where the agent training works. For now, I won't introduce the evaluation